Add Content Rules field to Configurations

eduardoj opened the pull request.

hennevogel commented on src/api/app/views/webui/code_of_conduct/index.html.haml:
```diff
@@ -0,0 +1,6 @@
+  .card-body
+    %h3.mb-4 Code of Conduct
+    %textarea.form-control.font-monospace.code-of-conduct{ disabled: true }
+      = @code_of_conduct
```
How about markdown? :)

eduardoj replied:
I switched it to markdown.

hennevogel replied:
Hm sorry  I don't think markdown is an option as I just remembered that `Configuration` has an API and markdown in XML is not possible AFAIR...

eduardoj replied:
In the end I changed it to use HTML, for consistency with the `description` field. Please take a look again.

hennevogel commented on src/api/app/views/layouts/webui/webui.html.haml:
```diff
@@ -65,11 +65,15 @@
+          - if @configuration.code_of_conduct.present?
+            .ms-2
+              = succeed '.' do
+                = link_to('Code of Conduct', code_of_conduct_index_path)
```
I think I would rather place this in `layouts/webui/footer` in the `Help` section.

eduardoj replied:
I didn't put it in the `Help` section because I didn't consider the "Code of Conduct" help...

hennevogel replied:
It kind of is, it's documentation on what is considered "good behavior" on this instance...

eduardoj replied:
I moved the link to the `Help` section.

eduardoj commented on src/api/.database_consistency.todo.yml:
```diff
@@ -492,6 +492,9 @@ Configuration:
       enabled: false
+  code_of_conduct:
+    LengthConstraintChecker:
+      enabled: false
```
Done.

I also added a commit, to validate the lenght of the rest of the fields of Configurations. Please take a look again.

danidoni commented on src/api/app/views/webui/code_of_conduct/_breadcrumb_items.html.haml:
Is this file needed?

eduardoj replied:
1) If we don't define it, we get this error:
```
ActionView::Template::Error (Missing partial webui/code_of_conduct/_breadcrumb_items, webui/webui/_breadcrumb_items with {:locale=>[:en], :formats=>[:html], :variants=>[], :handlers=>[:raw, :erb, :html, :builder, :ruby, :haml]}.

Searched in:
  * "/obs/src/api/app/views"
  * "/usr/lib64/ruby/gems/3.1.0/gems/view_component-3.5.0/app/views"
  * "/usr/lib64/ruby/gems/3.1.0/gems/kaminari-core-1.2.2/app/views"

Did you mean?  webui/user/breadcrumb_items
               webui/request/breadcrumb_items
               webui/session/breadcrumb_items
               webui/main/breadcrumb_items
               webui/users/breadcrumb_items
               webui/attribute/breadcrumb_items):
    1: %nav{ 'aria-label' => 'breadcrumb' }
    2:   %ol.breadcrumb.bg-transparent
    3:     = render partial: 'breadcrumb_items'
  
app/views/layouts/webui/_breadcrumbs.html.haml:3
app/views/layouts/webui/webui.html.haml:60
config/initializers/prefer_xml_over_html.rb:15:in `call'
```

2) Filling the breadcrumbs with a basic text like this:
```
%li.breadcrumb-item.active{ 'aria-current' => 'page' }
  Code Of Conduct
```
resulted in this rendering:

![Screenshot from 2023-09-18 11-27-59](https://github.com/openSUSE/open-build-service/assets/24919/b2bf09f4-48ac-4c2a-8303-6c0dd82bba92)


I found the repeated texts unnecessary redundant.

3) I personally preferred the actual option, wich results in:
![Screenshot from 2023-09-18 11-27-59](https://github.com/openSUSE/open-build-service/assets/24919/17f5cf7a-b0e2-40a0-9f97-9f5a9ebbd710)

danidoni replied:
Okay!

eduardoj commented:
@hennevogel, @danidoni, I added a commit to add the `code_of_conduct` field to the configurations API endpoints. I also addressed all your suggestions. Please take a look again.

hennevogel commented on src/api/app/views/webui/main/index.html.haml:
```diff
@@ -8,9 +8,7 @@
-        -# rubocop:disable Rails/OutputSafety
-        = raw @configuration['description']
-        -# rubocop:enable Rails/OutputSafety
+        = sanitize @configuration['description']
```
Why do you sanitize this now?

eduardoj replied:
To align behaviours on how we render similar fields.

eduardoj replied:
I see two options:
1. Use `sanitize @configuration[...]` in both files: `app/views/webui/main/index.html.haml` and `app/views/webui/code_of_conduct/index.html.haml`, or
2. Use `raw @configuration[...]`  in both files: `app/views/webui/main/index.html.haml` and `app/views/webui/code_of_conduct/index.html.haml`

Do you mean you prefer option _2._?

hennevogel replied:
> I prefer giving OBS admins safe and sane ways of modifying these fields.

Safe from themselves?

eduardoj replied:
> Safe from themselves?

Sure.

And I still can't imagine a valid counter case.

hennevogel replied:
Just because you can't imagine one doesn't mean it doesn't exist. But okay, we can sanitize if you want to.

hennevogel commented on src/api/app/controllers/webui/code_of_conduct_controller.rb:
```diff
@@ -0,0 +1,7 @@
+  def index
+    @code_of_conduct = ::Configuration.first.code_of_conduct
+
+    head :not_found if @code_of_conduct.blank?
```
If you just throw an `ActiveRecord::RecordNotFound` it'll do the right thing...

eduardoj replied:
Done. :+1:
